### DISCUSSION TITLE
COCOTB + Verilator reads stale signal values when on clock posedge

### DISCUSSION THREAD

**Discussion by HakamAtassi** [cocotb/cocotb#3877] [Q&A] [answered]
Hello!

I've been using cocotb for some time now and have run into the issue of cocotb fetching stale values of a signal on a posedge on several occasions. This issue typically goes something like:

> tests run fine for some time.
> some assertion fails.
> I check signal value against what cocotb claims it is using gtkwave
> signal matches expected, but cocotb is comparing the signal to its previous value

I have resorted to preforming sleep(0) at the top of my main simulation loop, which fixes things, but is less than ideal. 
Is this behavior a bug or is there a way to "serialize" the simulation prior to taking samples of the signals?

**Comment by ktbarrett** [comment, marked as answer]
This is how normal simulators and HDLs works. When you sensitize yourself to a value change on a signal, the process awakens after the value changes, but before any of the other sympathetic signals are updated (giving you the chance to add your own non-blocking writes for the next delta cycle).

If you want to read the signals after they have all "settled" you can use `await ReadOnly()` to put yourself in the postponed region. You won't be able to do any writes at this point. The other solution is to change your code to work more like HDL by sampling the last cycles settled values at the next clock edge.
